gslib/commands/iam.py

```diff
@@ -309,6 +316,55 @@ class IamCommand(Command):
       },
   )
 
+  def get_gcloud_storage_args(self):
+    sub_command = self.args.pop(0)
+    if sub_command == 'get':
+      gcloud_storage_map = GcloudStorageMap(gcloud_command=[
+          'alpha', 'storage', None, 'get-iam-policy', '--format=json'
+      ],
+                                            flag_map={})
+      if StorageUrlFromString(self.args[0]).IsObject():
+        gcloud_storage_map.gcloud_command[2] = 'objects'
+      else:
+        gcloud_storage_map.gcloud_command[2] = 'buckets'
+    elif sub_command == 'set':
+      gcloud_storage_map = GcloudStorageMap(
+          gcloud_command=[
+              'alpha', 'storage', None, 'set-iam-policy', '--format=json'
+          ],
+          flag_map={
+              '-a': GcloudStorageFlag('--all-versions'),
+              '-e': GcloudStorageFlag('--etag'),
+              # Hack to make gcloud recognize empty string value.
+              '_empty_etag': GcloudStorageFlag('--etag='),
+              '-f': GcloudStorageFlag('--continue-on-error'),
+              '-R': GcloudStorageFlag('--recursive'),
+              '-r': GcloudStorageFlag('--recursive'),
+          })
+      # Flags must be at the start of self.args to get parsed.
```

Comment:
No, this isn't referring to that bug. Flags must be at the start of `gsutil iam` b/c there's a variable number of cloud URLs. The structure is `gsutil iam set [FLAGS] [FILE URL] [CLOUD URLS...]`

If a flag isn't directly after "set" it will be incorrectly parsed as a URL.

```
$ gsutil iam set ~/Desktop/iam.json gs://bucket -r
gs://0noogler-kingdom/
file://-r
'_FileUrl' object has no attribute 'IsBucket'
$ gsutil iam set ~/Desktop/iam.json -r gs://bucket
file://-r
'_FileUrl' object has no attribute 'IsBucket'
```

Reply:
Got it, thanks!